### PR TITLE
WIP: feat(theme): Add dark mode

### PR DESCRIPTION
**Pull request by BrunnerLivio**
## Poll

**Use a reaction :+1: or :-1: on this PR to vote**, whether Dark Mode shall be implemented for docs.nestjs.com and you would happily use it!
 
Kamil:

> if there are more people interested in having the dark theme, I'm completely fine with it.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/docs.nestjs.com/blob/master/CONTRIBUTING.md

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] Other... Please describe:
```

## What is the current behavior?

Sunglasses

## What is the new behavior?

No Sunglasses

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

Dark Mode

I know from #613 @kamilmysliwiec is opposed to dark mode. Though since more and more platforms are migrating and I have been working the last year on a design system, I felt a huge urge to follow it as well ;)

Notes:

- This PR uses CSS variables. This means IE11 support will be dropped. Maybe double-check the analytics, whether this is something we still should support? FYI: Stackoverflow dropped the IE11 support as well, when they introduced dark mode with CSS variables.
- The NestJS primary color #ed2945 does not have a good contrast ratio with the background. I changed it to a lighter color (randomly picked). If you have a better strategy, feel free to let me know.

- Valor

- Companies do not look very good in dark mode Companies

Next steps:

- Consider refactoring graphs to [MermaidJS](https://mermaidjs.github.io/#/). Then we can easily convert those graphics to dark mode + images can be maintained in code.

### CODE REVIEW

**Comment by kamilmysliwiec** [comment]
Thanks for your contribution! Honestly, I'm not entirely convinced of this feature. I have a dark mode turned on in my OS (and I love it!), but I think dark themes in documentations make them less readable. I checked documentations of other popular projects like Angular, React or Vue, and I haven't noticed anything like this either. So what I'm thinking - if there are more people interested in having the dark theme, I'm completely fine with it. Nonetheless, it has to be disabled by default (for the sake of consistency between platforms) and available on-demand (so if you want to have it, you have to explicitly turn it on, otherwise, we should use the light theme by default).

**Comment by BrunnerLivio** [comment]
@kamilmysliwiec 
>  dark themes in documentations make them less readable
Before this feature would get merged, I would check it with Lighthouse to check the contrast colors (making it readable). Maybe automate it via #971 

> So what I'm thinking - if there are more people interested in having the dark theme, I'm completely fine with it. 

Makes sense, fine for me :)

> Nonetheless, it has to be disabled by default (for the sake of consistency between platforms) and available on-demand (so if you want to have it, you have to explicitly turn it on, otherwise, we should use the light theme by default).

Also if you have selected "prefer Dark Mode" as OS/Browser setting?

**Comment by kamilmysliwiec** [comment]
> Also if you have selected "prefer Dark Mode" as OS/Browser setting?

Yes, following other documentations and StackOverflow. We can also provide a third option, "Light, Dark, System" (inspired by StackOverflow).

**Comment by bradtaniguchi** [comment]
> I checked documentations of other popular projects like Angular, React or Vue, and I haven't noticed anything like this either.

Docker docs have the "toggable" dark mode feature: https://docs.docker.com/
The doc's "light" color theme is still that light-ish blue, so logos that are white are still readable in both dark and light. Some icons change color depending on the theme, but generally the docs have limited colors and images.

As someone that uses **dark theme on basically everything**, either via native or via some extension/styling mod. There are a bunch of cases where going dark theme can screw stuff up, like images become unreadable. I'd say the docs are totally in the realm of possibility of having a dark theme, and Angular makes toggle themes pretty easy. (or at least I think so hehe)

**Comment by jmcdo29** [comment]
@BrunnerLivio @kamilmysliwiec about a year later, and it seems there is some want from the community around dark mode. What would we need to do to scope this out and merge it?

**Comment by MSandeep96** [comment]
Just wanted to add that Redux and Redux toolkit docs have a toggle for dark mode. People who work at night would definitely appreciate this feature.

**Comment by BrunnerLivio** [comment]
@jmcdo29 @kamilmysliwiec I’d love to follow-up on this one! Just let me know when I can proceed rebasing!

**Comment by Haltarys** [comment]
I personally use the [DarkReader](https://darkreader.org/) extension, which works almost perfectly with any website (including the docs of NestJS).

**Comment by BrunnerLivio** [comment]
Rebased & dark mode toggle added 🦜

**Comment by MSandeep96** [comment]
Bump?

**Comment by chris9740** [comment]
> Thanks for your contribution! Honestly, I'm not entirely convinced of this feature. I have a dark mode turned on in my OS (and I love it!), but I think dark themes in documentations make them less readable. I checked documentations of other popular projects like Angular, React or Vue, and I haven't noticed anything like this either. So what I'm thinking - if there are more people interested in having the dark theme, I'm completely fine with it. Nonetheless, it has to be disabled by default (for the sake of consistency between platforms) and available on-demand (so if you want to have it, you have to explicitly turn it on, otherwise, we should use the light theme by default).

I agree that a lot of people will have an easier time reading the documentation in light-theme, however; I personally find it hard to read websites with a light background because of my visual impairment (I have moderately severe astigmatism), and dark themes are a lot less straining on my eyes (this also applies to some people outside the context of eye conditions). I wouldn't mind light theme being the default, and having the option to turn on dark theme would greatly improve my experience on the NestJS documentation website

**Review by jmcdo29 — APPROVED** [review]

**Comment by hichemfantar** [comment]
Any plans to merge this?

File: src/app/homepage/dark-mode-toggle/dark-mode-toggle.component.ts
Comment (Exac) [inline src/app/homepage/dark-mode-toggle/dark-mode-toggle.component.ts:29]: Using the Media Matcher instead of `window` makes it easier to use Universal.

https://material.angular.io/cdk/layout/overview#mediamatcher

**Review by Exac — CHANGES_REQUESTED** [review]

**Comment by armujahid** [comment]
Docs work fine with [darkreader](https://darkreader.org/) extension without any noticeable issue
image

**Comment by Tony133** [comment]
Hi everyone 👋🏻 ,

I made some updates to this PR, fixed tons of conflicts, made some updates and changes you colors. 
To avoid creating another PR since there has been a long discussion here I have improved this PR, hoping it will be merged.

I attach a screenshot of what it looks like now:

screen-dark-mode





Here is the link for a preview: https://deploy-preview-1138--docs-nestjs.netlify.app/  🔥🚀

so you can view it.  

If you are interesting and want this PR to be merged approve it! 🙏🏻

I hope I have been helpful.

@BrunnerLivio  mission accomplished! ✅

Dark mode for NestJS documentation 👀 🔥, the story continues..... 🔥 🚀

P.s. the CI fails but not because of these changes ( for info see here: https://github.com/nestjs/docs.nestjs.com/actions/runs/5524617442/jobs/10077158987?pr=1138)

**Review by Tony133 — APPROVED** [review]

**Review by micalevisk — APPROVED** [review]

**Comment by kamilmysliwiec** [comment]
@Tony133 great job!

I believe we could also apply CSS filters to template illustrations we have, for example, the one we have on the Modules page:

image

to:

image


CSS:
```css
filter: invert(1) contrast(0.85);
box-shadow: 0 0 50px #fff;
```

**Comment by kamilmysliwiec** [comment]
Because migrating something that has proved to work well for several years makes zero sense

**Comment by Tony133** [comment]
Ok, I added the css rule for illustrative images 🔥, these are the following links to check them:

- https://deploy-preview-1138--docs-nestjs.netlify.app/controllers
- https://deploy-preview-1138--docs-nestjs.netlify.app/providers
- https://deploy-preview-1138--docs-nestjs.netlify.app/modules
- https://deploy-preview-1138--docs-nestjs.netlify.app/middleware
- https://deploy-preview-1138--docs-nestjs.netlify.app/exception-filters
- https://deploy-preview-1138--docs-nestjs.netlify.app/pipes
- https://deploy-preview-1138--docs-nestjs.netlify.app/guards
- https://deploy-preview-1138--docs-nestjs.netlify.app/interceptors
- https://deploy-preview-1138--docs-nestjs.netlify.app/microservices/basics
- https://deploy-preview-1138--docs-nestjs.netlify.app/microservices/redis
- https://deploy-preview-1138--docs-nestjs.netlify.app/websockets/gateways
- https://deploy-preview-1138--docs-nestjs.netlify.app/fundamentals/lifecycle-events


There are these two pages:

- https://deploy-preview-1138--docs-nestjs.netlify.app/openapi/types-and-parameters
- https://deploy-preview-1138--docs-nestjs.netlify.app/recipes/documentation

which have images with always white background, but the css rule is not very suitable in this case 😅. (See the example screenshot) and at the moment I left them as they are without using the css rule. 😅

screen-dark

screen-dark-2


Let me know what you think and if you have any ideas for those two pages. 😄

File: src/app/shared/components/dark-mode-toggle/dark-mode-toggle.component.ts
Comment (kamilmysliwiec) [inline src/app/shared/components/dark-mode-toggle/dark-mode-toggle.component.ts:27]: This logic doesn't seem to work as expected (explicit setting is not memorized)

File: src/app/shared/components/dark-mode-toggle/dark-mode-toggle.component.ts
Comment (kamilmysliwiec) [inline src/app/shared/components/dark-mode-toggle/dark-mode-toggle.component.ts:22]: Let's use the current (light) theme by default, regardless of the media value

Comment (Tony133) [reply src/app/shared/components/dark-mode-toggle/dark-mode-toggle.component.ts]: Done! ✅ ,now by default is in light mode.

link-preview: https://deploy-preview-1138--docs-nestjs.netlify.app

**Comment by kamilmysliwiec** [comment]
Button colors don't match

image

image

**Comment by kamilmysliwiec** [comment]
It doesn't match the logo color now though

**Comment by Tony133** [comment]
> It doesn't match the logo color now though

Okay, I would say it should now be similar/equal to the color of the logo  😅
Screenshot 2023-07-13 alle 18 15 15

link preview: https://deploy-preview-1138--docs-nestjs.netlify.app/ 🔥

File: src/app/shared/components/dark-mode-toggle/dark-mode-toggle.component.ts
Comment (Tony133) [reply src/app/shared/components/dark-mode-toggle/dark-mode-toggle.component.ts]: On this part of the logic to me it seems to work correctly, except renaming the names, I left everything as Livio had configured 3 years ago. 😅 Let me know what you mean about this part. 🙂

**Review by Exac — COMMENTED** [review]
valor-logo-light-selected
We should use this SVG to remove the problematic white artifacts `https://valor-software.com/assets/img/valor-img/valor-logo.svg`.

**Comment by Exac** [comment]
After signing up for the newsletter, the checkbox is still quite light:
image

**Comment by Tony133** [comment]
> After signing up for the newsletter, the checkbox is still quite light: image

Fixed! ✅ see commit here: https://github.com/nestjs/docs.nestjs.com/pull/1138/commits/319555e735a307e78dc23c61db6873a64b7429d7

Thanks for the reporting!

Link preview: https://deploy-preview-1138--docs-nestjs.netlify.app/  🔥

File: src/app/shared/components/theme-mode-toggle/theme-mode-toggle.component.html
Comment (Exac) [inline src/app/shared/components/theme-mode-toggle/theme-mode-toggle.component.html:4]: Nitpick: Should the button show the current state, or the state that will be applied when it is toggled?

**Review by Exac — APPROVED** [review]

File: src/app/shared/components/theme-mode-toggle/theme-mode-toggle.component.html
Comment (Tony133) [reply src/app/shared/components/theme-mode-toggle/theme-mode-toggle.component.html:4]: > Nitpick: il pulsante dovrebbe mostrare lo stato corrente o lo stato che verrà applicato quando viene attivato o disattivato?

The button shows the current state of how the theme is configured, so clicking the button changes the state to either "dark" or "light" mode, while also changing the icon. By default it is set to light mode at Kamil's request (see previous comments).🙂

Comment (Exac) [reply src/app/shared/components/theme-mode-toggle/theme-mode-toggle.component.html:4]: Sure, what I mean essentially is that when a user opens the docs, and wants to switch to dark mode, they might search the page for the button to turn on dark mode. The button will show a sun, but the page is already in light mode.

I'm sure users will get it, I don't think there is any standard on whether to show the current state, or the opposite state on a button.

I think the current approach works, and I think we are ready to merge this PR.

Comment (Tony133) [reply src/app/shared/components/theme-mode-toggle/theme-mode-toggle.component.html:4]: I hope too that it will be merged soon 🙏🏻, it is a nice feature for documentation then especially in the evening it is very useful if you want to read the documentation of NestJS 😻 😎

**Comment by j-tag** [comment]
@kamilmysliwiec please consider merging this feature. There are many people who are [photophobic](https://en.wikipedia.org/wiki/Photophobia). This is a disorder and is very frustrating. For many people this is not about style or being cool, it's just about health.

**Comment by BrunnerLivio** [comment]
Dark mode is merged? Woop woop 🎉 

undertaker-wake-up

**Comment by gterras** [comment]
There is no button to switch back to the light theme ? That's a regression not a feature @BrunnerLivio  @kamilmysliwiec

**Comment by kamilmysliwiec** [comment]
Just click on the icon (moon):

image

Also, theme remains light by default anyways unless you manually switch to the dark one

**Comment by gterras** [comment]
> Just click on the icon (moon):

I had no icons there: 

image

After investigation it appears that this whole bar is blocked by uBlock Origin (default rules), maybe because of its `social-wrapper` class? Theme button should probably be in another semantic element anyway. 

> Also, theme remains light by default anyways unless you manually switch to the dark one

Not my case here, it went dark automatically after a refresh (firefox/linux). Note that my OS theme is set to dark but Firefox settings is set to Light, I don't know if these are used somewhere here.

**Comment by kamilmysliwiec** [comment]
This should be fixed now @gterras

**Comment by gterras** [comment]
> This should be fixed now @gterras

Thanks I can confirm both issues are fixed.

**Pull request merged**
